src/Framework/FileClassifier.cs

```diff
@@ -0,0 +1,196 @@
+    {
+        /// <summary>
+        ///     StringComparison used for comparing paths on current OS.
+        /// </summary>
+        /// <remarks>
+        ///     TODO: Replace RuntimeInformation.IsOSPlatform(OSPlatform.Linux) by NativeMethodsShared.OSUsesCaseSensitivePaths once it is moved out from Shared
+        /// </remarks>
+        private static readonly StringComparison PathComparison = RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
+
+        /// <summary>
+        ///     Single, static instance of an array that contains a semi-colon ';', which is used to split strings.
+        /// </summary>
+        private static readonly char[] s_semicolonDelimiter = {';'};
+
+        /// <summary>
+        ///     Singe, static <see cref="Lazy{T}"/> instance of shared file FileClassifier for <see cref="Shared"/> member.
+        /// </summary>
+        private static readonly Lazy<FileClassifier> s_sharedInstance = new(() => new FileClassifier());
+
+        /// <summary>
+        ///     Serves purpose of thread safe set of known immutable directories.
+        /// </summary>
+        /// <remarks>
+        ///     Although <see cref="ConcurrentDictionary{TKey,TValue}"></see> is not memory wise optimal solution, int this particular case it does no matter
+        ///     as much as expected size of this set is ~5 and in very extreme cases less then 100.
+        /// </remarks>
+        private readonly ConcurrentDictionary<string, string> _knownImmutableDirectories = new();
+
+        /// <summary>
+        ///     Copy on write snapshot of <see cref="_knownImmutableDirectories"/>.
+        /// </summary>
+        private IReadOnlyList<string> _knownImmutableDirectoriesSnapshot = Array.Empty<string>();
+
+        /// <summary>
+        ///     Creates default FileClassifier which following immutable folders:
+        ///     Classifications provided are:
+        ///     <list type="number">
+        ///         <item>Program Files</item>
+        ///         <item>Program Files (x86)</item>
+        ///         <item>Default .nuget cache location</item>
+        ///         <item>Visual Studio installation root</item>
+        ///     </list>
+        /// </summary>
+        /// <remarks>
+        ///     Individual projects NuGet folders are added during project build by calling
+        ///     <see cref="RegisterImmutableDirectories" />
+        /// </remarks>
+        public FileClassifier()
+        {
+            string? programFiles32 = Environment.GetEnvironmentVariable("ProgramFiles(x86)");
+            string? programFiles64 = Environment.GetEnvironmentVariable("ProgramW6432");
+
+            if (!string.IsNullOrEmpty(programFiles32))
+            {
+                RegisterImmutableDirectory(Path.Combine(programFiles32, "Reference Assemblies", "Microsoft"));
+                RegisterImmutableDirectory(Path.Combine(programFiles32, "dotnet"));
```

Comment:
Should we have these if they're _not_ `$(NetCoreRoot)`?

Reply:
It is attempt to put that on known list of immutable dirs so even code which runs before is that property resolved would leverage from this. I am not sure however, how much code that is. My gut feelings is 'not a lot'.
Shall we keep it here, or just relay of that one from `RequestBuilder`?

Reply:
I don't feel strongly but would propose: do just the other one and if we observe it being insufficient we can do it here.